### ISSUE TITLE
Understanding overhead showing in Chrome trace

### ISSUE THREAD

**Issue by kastiglione** [bazelbuild/bazel#10702]
### Description of the problem  request:

We have a long final action at the end of our build that we'd like to speed up. When looking at a Chrome trace of our build, that final action has much more overhead than other actions in the build.

![Screen Shot 2020-02-03 at 2 43 37 PM](https://user-images.githubusercontent.com/43429/73697513-990b1a80-4693-11ea-86f5-ac92d7554b09.png)

My reading of this is that the action (`BundleTreeApp` a python script in rules_apple) is ~6s, and then after that Bazel does some post-processing in `ActionContinuation.execute` and then yet more post-processing in the top level invocation (skyframe?).

 In the above screenshot, `ActionContinuation.execute` appears to have 2s of post-processing "overhead", and "Bundling, processing and signing" has another 2s of post-processing "overhead".
 
### Feature requests: what underlying problem are you trying to solve with this feature?

When looking at a Chrome trace: How can we understand how and why Bazel overhead exists, and what can be done about it.

We see overhead at the beginning of a build as well, but this issue is focused on how to understand performance factors for poorly performing actions like this.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

You'd need a sufficiently large iOS project that has a non-trivial `.app` bundle.

### What operating system are you running Bazel on?

macOS

### What's the output of `bazel info release`?

release 2.0.0

**Comment by ghvg1313** [comment]
Don't know if this is related, but we are also experiencing similar overhead when calling `build` and `test` action separately (`bazel build <target>` then `bazel test <target>`) with tree-artifacts turned on.

Traces:
[test trace with tree-artifacts turned off](https://github.com/bazelbuild/bazel/files/4162236/no-tree.trace.gz)
[test trace with tree-artifacts turned on](https://github.com/bazelbuild/bazel/files/4162239/tree_artifacts_test.trace.gz)

The overhead is not too bad considering this is only one target, but you can clearly see multiple `action dependency checking` distributed across threads before test runner executes, which, you don't see when tree artifacts turned off.
The performance of dependency checking gets significantly worse when there're multiple targets under test

**Comment by brentleyjones** [comment]
Looking into this, it seems the post processing is Bazel collecting the contents of the `declare_directory`?

**Comment by BalestraPatrick** [comment]
This is still a valid issue. Our iOS app bundle with 16K files takes over 75s to complete this action, even though the rules_apple `BundleTreeApp` action was optimized to only take 10s, we have no easy way to optimize the rest of time that Bazel spends in collecting the contents of all directories and files.
<img width="1367" alt="Screenshot 2021-12-16 at 10 23 03 PM" src="https://user-images.githubusercontent.com/3658887/146451084-cc0a8cd7-d063-4c02-a8b9-b43c7fb52127.png">

**Comment by chiragramani** [comment]
Yeah, we are seeing this issue as well.

![image](https://user-images.githubusercontent.com/11925399/193376366-be2f2612-ae2d-4a6a-ae41-83592523c6dd.png)

- Linking took ~8.5 seconds but actuallyCompleteAction took 6.4 seconds. (this makes the overall linking action take ~15 seconds) 
- Symlink creation takes significant time. (~12.5 seconds)
![image](https://user-images.githubusercontent.com/11925399/193376396-5a96eecc-4f3d-4055-8a8e-82ec38c85e8f.png)

Any optimizations here would give us really good wins, especially for the incremental builds.

**Comment by brentleyjones** [comment]
```
git clone https://github.com/buildbuddy-io/rules_xcodeproj.git
cd rules_xcodeproj
git checkout bj/overhead-demo
cd examples/integration

bazel build //iOSApp
```

You can then look at the `Linking iOSApp/Source/iOSApp_bin`, `Creating symlink bazel-out/applebin_ios-ios_sim_arm64-fastbuild-ST-4e6c2a19403f/bin/iOSApp/Source/iOSApp_lipobin`, and `Creating symlink bazel-out/applebin_ios-ios_sim_arm64-fastbuild-ST-4e6c2a19403f/bin/iOSApp/Source/iOSApp-intermediates/iOSApp_ExecutableName` actions:

<img width="902" alt="CleanShot 2022-10-06 at 09 41 57@2x" src="https://user-images.githubusercontent.com/158658/194343198-389ccd2b-a3f4-4302-907e-64e904e34afc.png">

The blank space at the end of all of them scale with the size of the linked binary (so this example will be _milliseconds_, 73 in my case on each action, but on real projects it grows to multiple seconds, and is unnecessary overhead).

**Comment by wilwell** [comment]
Thanks @brentleyjones for reproduction code.
I was able to reproduce your example only  on M1 machine.
I investigated `Linking iOSApp...` action. On that case the difference between `ActionContinuation.execute` and `subprocess.run` is `sandbox.delete` and `sandbox.CopyOutputs`, which is reasonable. 

But `actuallyCompleteAction` following  `ActionContinuation.execute` is more interesting. I'm not sure what exactly going there, but I suppose that this is some manipulations with output files, e.g. [flush of ActionFileSysyem](https://github.com/bazelbuild/bazel/blob/1620b17ccfc1167e66c4888488d272ad1a2022f4/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java#L1204).


The profile is [here](https://github.com/bazelbuild/bazel/files/9829693/profile.gz).

**Comment by brentleyjones** [comment]
My concern is more the symlinks actions taking the same amount of extra time. Those shouldn't be sandbox copy and delete times, since it's just a symlinks. So I agree, the more concerning part isn't the sandbox but the `actuallyComplete` portion.

**Comment by chiragramani** [comment]
Came across this issue - https://github.com/bazelbuild/bazel/issues/14125 that also talks about some overhead involved with`actuallyCompleteAction`
```
Creating the output takes .75s, but postprocessing (checksumming?) via actuallyCompleteAction takes 5s
```

**Comment by tjgq** [comment]
Sadly, `ctx.actions.symlink` with a `declare_file` or `declare_directory` output parameter doesn't actually optimize out the copy. The action itself does create a symlink in the filesystem, but Skyframe dereferences it and loses track of the fact that it was a symlink when the action outputs are collected (which explains the additional checksumming and copying overhead).

One possible solution is to generalize 32b0f5a to also keep track of symlinks in Skyframe for the local execution case. I know that this isn't trivial to do because I tried (and failed) to do that in an earlier iteration of that CL.

**Comment by tjgq** [comment]
I'm going to close this as a duplicate of #14125.

**Comment by larsrc-google** [comment]
If you don't already, set `--reuse_sandbox_directories` to cut down on the sandboxing overhead.

**Comment by brentleyjones** [comment]
@larsrc-google Should the default on that be changed?

**Comment by larsrc-google** [comment]
There has been a problem with what seems to be bad interactions with async deletes that has made me hold off on flipping it.

**Comment by brentleyjones** [comment]
Would those errors look like this 😅?
> ERROR: /Users/brentley/Developer/rules_xcodeproj/tools/generator/BUILD:20:14: Linking tools/generator/libgenerator.library.a [for tool] failed: I/O exception during sandboxed execution: /Users/brentley/Developer/rules_xcodeproj/bazel-output-base/rules_xcodeproj/build_output_base/sandbox/darwin-sandbox/43/execroot/com_github_buildbuddy_io_rules_xcodeproj/bazel-out (No such file or directory)

**Comment by larsrc-google** [comment]
Yes, that looks similar. I've spent some time looking into it but not found a viable hypothesis yet.

**Comment by brentleyjones** [comment]
I could very easily reproduce it. Would a repro be of value to you?

**Comment by larsrc-google** [comment]
Yes please, I've had trouble getting a stable repro.
